File: src/69.ts
import primes from './core/primes';

// Computes the value of n for which n/φ(n) is maximized.
//
// @remarks
// It's not necessary to actually compute the totient function for each number.  The ratio will be largest with the
// denominator is as small as possible.  This will occur when n has many prime divisors.
//
// See {@link docs/problem-69.md}.
// See {@link https://projecteuler.net/problem=69}.
export default function compute(limit: number) {
  let ps = primes();
  let acc = 1;

  while (ps.length > 0) {
    const n = acc * ps[0];
    if (n > limit) {
      return acc;
    }

    acc = n;
    ps = ps.slice(1);
  }

  throw new Error('not enough primes');
}

File: project_euler/69.py
# Problem 69: Totient Maximum
#
# Find the value of n ≤ 1_000_000 for which n/φ(n) is a maximum.
#
# See https://projecteuler.net/problem=69
from typing import cast
from sympy import primerange


def run(limit=1_000_000) -> int:
    acc = 1

    for x in primerange(2, limit):
        p = cast(int, x)
        n = acc * p
        if n > limit:
            return acc
        acc = n

    raise ValueError("no solution")


def test_run():
    assert run() == 510510
